sklearn/tree/_criterion.pyx

```diff
@@ -586,6 +586,11 @@ cdef class ClassificationCriterion(Criterion):
         elif dest[0] > upper_bound:
             dest[0] = upper_bound
 
+        if dest[0] < 0.:
+            dest[0] = 0
+        elif dest[0] > 1.:
+            dest[0] = 1.
```

Comment:
This forces the support of values to be in $[0, 1]$, but is this actually the appropriate implementation? Isn't the problem else where in the implementation?

Reply:
I don't know. I was trying to understand what the lower and upper bound that are stored in the tree builder stack (and then passed to this function) are used for/how their value is computed but not sure.

The very first node that gets pushed to the stack uses `-inf` and `+inf` for the lower and upper bounds. Which would suggest it is possible for the values to be out side the range of zero to one?

Reply:
You were right @jjerphan the first attempt was a quick and dirty fix. As suspected there was something else going on... Now with [d74b012](https://github.com/scikit-learn/scikit-learn/pull/27639/commits/d74b0127521acc9a3e15f0556d0cb49991bc785a) we're closer to a real solution, although we still have to understand the remaining floating point precisions issues...

Reply:
I think clipping this is a good hot fix for the support and we may first proceed with it.

Reply:
But values are not supposed to be in [0, 1].

Reply:
> But values are not supposed to be in [0, 1].

Exactly ! I was not aware of this until your comment https://github.com/scikit-learn/scikit-learn/pull/27639#discussion_r1376313430 @glemaitre.